Runtime/Services/BacktraceSession.cs

```diff
@@ -0,0 +1,354 @@
+
+        /// <summary>
+        /// Lock object
+        /// </summary>
+        private object _object = new object();
+
+        private float _lastUpdateInvoke = 0;
+        private readonly Stack<SessionSubmissionJob> _submissionJobs = new Stack<SessionSubmissionJob>();
+
+        /// <summary>
+        /// Create new Backtrace session instance
+        /// </summary>
+        /// <param name="attributeProvider">Backtrace client attribute provider</param>
+        /// <param name="uploadUrl">Upload URL</param>
+        /// <param name="timeIntervalInMs">Update time interval in MS</param>
+        /// <param name="maximumNumberOfEventsInStore">Determine how many events we can store in event store</param>
+        public BacktraceSession(
+            AttributeProvider attributeProvider,
+            string uploadUrl,
+            long timeIntervalInMs,
+            uint maximumNumberOfEventsInStore)
+        {
+            SubmissionUrl = uploadUrl;
+            _attributeProvider = attributeProvider;
+            _maximumNumberOfEventsInStore = maximumNumberOfEventsInStore;
+            _timeIntervalInMs = timeIntervalInMs;
+            RequestHandler = new BacktraceHttpClient();
+        }
+
+        /// <summary>
+        /// Send startup event ot Backtrace
+        /// </summary>
+        public void SendStartupEvent()
+        {
+            Send(new UniqueEvent[0], new SessionEvent[1] { new SessionEvent(StartupEventName) }, DefaultNumberOfRetries);
+        }
+
+        /// <summary>
+        /// Backtrace session update interval method used by Backtrace client to update session events time
+        /// </summary>
+        /// <param name="time">Current game time</param>
+        public void Tick(float time)
+        {
+            _lastUpdateInvoke = time;
+            lock (_object)
+            {
+                // check submission job
+                while (_submissionJobs.Count != 0 && _submissionJobs.First().NextInvokeTime < time)
+                {
+                    var submissionJob = _submissionJobs.Pop();
+                    Send(submissionJob.UniqueEvents, submissionJob.SessionEvents, submissionJob.NumberOfRetries);
+                }
+            }
+            if (_timeIntervalInMs == 0)
+            {
+                return;
+            }
+            lock (_object)
+            {
+
+                var intervalUpdate = (time - _lastUpdateTime) >= _timeIntervalInMs;
+                var reachedEventLimit = _maximumNumberOfEventsInStore == Count() && _maximumNumberOfEventsInStore != 0;
+                if (intervalUpdate == false && reachedEventLimit == false)
+                {
+                    // nothing more to update
+                    return;
+                }
+                _lastUpdateTime = time;
+            }
+            Send();
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace
+        /// </summary>
+        public void Send()

```

Comment:
I'm not a fan of how we're doing recursion by overloading a function of the same name 3 times.  I'd consider giving these unique names.

Reply:
How about renaming `Send` method with unique/session events parameter to `SendPayload`?